ruby/examples/lib/register_user.rb

```diff
@@ -0,0 +1,22 @@
+class RegisterUser
+  Request = Struct.new(:email, :name, :password)
+  Response = Struct.new(:values, :errors)
+
+  attr_reader :gateway
+
+  def initialize(gateway: Gateway.new)
+    @gateway = gateway
+  end
+
+  def register(request)
+    errors = gateway.create_user(request.to_h)
+    Response.new(request.to_h, errors)
+  end
+
+  class Gateway
+    def create_user(attributes)
```

Comment:
Don't like how the attributes are passed straight through to the gateway here. Creates some obscurity about what the shape of the message accepted by the Gateway is. 

I prefer explicit passing of a reusable "User" domain object. Then you can hang behaviours related to the User off the that Domain object.